tree/dataframe/src/RSqliteDS.cxx

```diff
@@ -0,0 +1,260 @@
+
+#include <algorithm>
+#include <cstring>
+
+namespace ROOT {
+
+namespace RDF {
+
+RSqliteDS::RSqliteDS(std::string_view fileName, std::string_view query)
+   : fDb(NULL)
+   , fQuery(NULL)
+   , fNSlots(1)
+   , fNRow(0)
+{
+   int retval;
+
+   retval = sqlite3_open_v2(std::string(fileName).c_str(), &fDb,
+     SQLITE_OPEN_READONLY | SQLITE_OPEN_NOMUTEX, NULL);
+   if (retval != SQLITE_OK) SqliteError(retval);
+
+   retval = sqlite3_prepare_v2(fDb, std::string(query).c_str(), -1, &fQuery, NULL);
+   if (retval != SQLITE_OK) SqliteError(retval);
+
+   int colCount = sqlite3_column_count(fQuery);
+   retval = sqlite3_step(fQuery);
+   if ((retval != SQLITE_ROW) && (retval != SQLITE_DONE)) SqliteError(retval);
+
+   fValues.resize(colCount);
+   for (int i = 0; i < colCount; ++i) {
+      fColumnNames.emplace_back(sqlite3_column_name(fQuery, i));
+      int type = SQLITE_NULL;
+      // Try first with the declared column type and then with the dynamic type
+      // for expressions
+      const char *declTypeCstr = sqlite3_column_decltype(fQuery, i);
+      if (declTypeCstr == NULL) {
+         if (retval == SQLITE_ROW) type = sqlite3_column_type(fQuery, i);
+      } else {
+         std::string declType(declTypeCstr);
+         if (declType == "INTEGER") type = SQLITE_INTEGER;
+         else if (declType == "FLOAT") type = SQLITE_FLOAT;
+         else if (declType == "TEXT") type = SQLITE_TEXT;
+         else if (declType == "BLOB") type = SQLITE_BLOB;
+         else throw std::runtime_error("Unexpected column decl type");
+      }
+
+      switch (type) {
+      case SQLITE_INTEGER:
+         fColumnTypes.push_back(Types::kInteger);
+         fValues[i].fType = Types::kInteger;
+         fValues[i].fPtr = &fValues[i].fInteger;
+         break;
+      case SQLITE_FLOAT:
+         fColumnTypes.push_back(Types::kReal);
+         fValues[i].fType = Types::kReal;
+         fValues[i].fPtr = &fValues[i].fReal;
+         break;
+      case SQLITE_TEXT:
+         fColumnTypes.push_back(Types::kText);
+         fValues[i].fType = Types::kText;
+         fValues[i].fPtr = &fValues[i].fText;
+         break;
+      case SQLITE_BLOB:
+         fColumnTypes.push_back(Types::kBlob);
+         fValues[i].fType = Types::kBlob;
+         fValues[i].fPtr = &fValues[i].fBlob;
+         break;
+      case SQLITE_NULL:
+         // TODO: Null values in first rows are not well handled
+         fColumnTypes.push_back(Types::kNull);
+         fValues[i].fType = Types::kNull;
+         fValues[i].fPtr = &fValues[i].fNull;
+         break;
+      default:
+         throw std::runtime_error("Unhandled data type");
+      }
+   }
+
+   fTypeNames[Types::kInteger] = "Long64_t";
+   fTypeNames[Types::kReal] = "double";
+   fTypeNames[Types::kText] = "std::string";
+   fTypeNames[Types::kBlob] = "std::vector<unsigned char>";
+   fTypeNames[Types::kNull] = "void *";
+}
+
+
+RSqliteDS::~RSqliteDS()
+{
+   sqlite3_finalize(fQuery);
+   sqlite3_close_v2(fDb);
```

Comment:
`sqlite3_finalize` returns the error code of the most recent operation on `fQuery`. Closing can possibly fail with `SQLITE_BUSY`, in which case resources are leaked. This should not happen the way it is used in this class because we cleanup the prepared statement before.

Reply:
maaaaybe if you want you could add a comment inside the method with this answer, just to prevent other ROOT devs to ask themselves the same question

Reply:
That reminds me that the coding convention requires at least one line of documentation per functions.

Reply:
Dropping a few more comments is a good idea.  Let me extend the pull request. @pcanal Do you typically comment all methods or just the public ones?

Reply:
The rule applies to all.  Albeit I would see it is necessary only for public one *and* for private one whose name is not obviously describing the content and its parameter.

Reply:
> i didn't know this...and I cannot find any mention of this e.g. here?

Good point.  This is missing on that link.   But it is definitively the rule (note for example the number of destructor with the 'obvious' one liner :) ).

Reply:
@pcanal it's unwritten and unenforced -- if nothing else, if it's a rule, it should be stated in the new coding conventions in #2298

Reply:
I updated #2298 according by updating directly @Axel-Naumann 's branch.